source/opt/types.h

```diff
@@ -205,6 +194,34 @@ class Type {
   DeclareCastMethod(RayQueryKHR)
 #undef DeclareCastMethod
 
+protected:
+  // Add any type-specific state to |hash| and returns new hash.
+  virtual size_t ComputeExtraStateHash(size_t hash, SeenTypes* seen) const = 0;
+
+  // helpers for incrementally computing hash value
+  // http://open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3876.pdf
+  template <typename T>
+  static size_t hash_combine(std::size_t seed, const T& val) {
```

Comment:
Could you make these utility functions in "source/util"?  They do not seem to be type specific.